### PR TITLE
controller-grpc: gRPC server/client WIP

### PR DESCRIPTION
**Pull request by jvatic**

### CODE REVIEW

File: controller-grpc/controller.proto [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+message StreamEventsRequest {
+  string app_id = 1;
+  repeated string object_types = 2; // []ct.EventType
+  string object_id = 3;
+  bool post = 4;
+  int32 count = 5;
+}
+
+message App {
+  string id = 1;
```
Comment (titanous) [inline controller-grpc/controller.proto]: This should be `name` as explained above.

File: controller-grpc/controller.proto [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  string name = 2;
```
Comment (titanous) [inline controller-grpc/controller.proto]: This should be `display_name` as explained above.

File: controller-grpc/controller.proto [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+  map<string, string> meta = 3;
```
Comment (titanous) [inline controller-grpc/controller.proto]: This should be called `labels`.

File: controller-grpc/controller.proto [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+  google.protobuf.Timestamp created_at = 4;
+  google.protobuf.Timestamp updated_at = 5;
```
Comment (titanous) [inline controller-grpc/controller.proto]: `create_time` / `update_time`

File: controller-grpc/controller.proto [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+  int32 deploy_timeout = 6;
+  string strategy = 7;
+  Release release = 8;
```
Comment (titanous) [inline controller-grpc/controller.proto]: Should this `release` be here?

File: controller-grpc/controller.proto [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+  Release release = 2;
+}
+
+message CreateDeploymentRequest {
+  string app_id = 1;
+  string release_id = 2;
+}
+
+message StreamEventsRequest {
+  string app_id = 1;
```
Comment (titanous) [inline controller-grpc/controller.proto]: This should be called `parent` as explained.

File: controller-grpc/controller.proto [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+  string strategy = 7;
+  Release release = 8;
+}
+
+message Release {
+  string id = 1;
+  App app = 2;
+  repeated string artifact_ids = 3;
+  map<string, string> env = 4;
+  map<string, string> meta = 5;
```
Comment (titanous) [inline controller-grpc/controller.proto]: `labels`

File: controller-grpc/controller.proto [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+  map<string, ProcessType> processes = 6;
+  google.protobuf.Timestamp created_at = 7;
```
Comment (titanous) [inline controller-grpc/controller.proto]: `create_time`

File: controller-grpc/controller.proto [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+  google.protobuf.Timestamp created_at = 4;
+  google.protobuf.Timestamp updated_at = 5;
+  int32 deploy_timeout = 6;
+  string strategy = 7;
+  Release release = 8;
+}
+
+message Release {
+  string id = 1;
+  App app = 2;
```
Comment (titanous) [inline controller-grpc/controller.proto]: I don't think this `app` should be here?

File: controller-grpc/controller.proto [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+  map<string, string> meta = 3;
+  google.protobuf.Timestamp created_at = 4;
+  google.protobuf.Timestamp updated_at = 5;
+  int32 deploy_timeout = 6;
+  string strategy = 7;
+  Release release = 8;
+}
+
+message Release {
+  string id = 1;
```
Comment (titanous) [inline controller-grpc/controller.proto]: `name`

File: controller-grpc/controller.proto [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+  App app = 2;
+  repeated string artifact_ids = 3;
```
Comment (titanous) [inline controller-grpc/controller.proto]: `artifact_names`

File: controller-grpc/controller.proto [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+
+message StreamAppLogRequest {
+  string app_id = 1; // can be either id or name
+  LogAggregatorLogOpts opts = 2;
+}
+
+message CreateReleaseRequest {
+  string app_id = 1;
+  Release release = 2;
+}
```
Comment (titanous) [inline controller-grpc/controller.proto]: This should have a `request_id` to enforce idempotence.

**Comment by jvatic** [comment]
@titanous ready for another review

**Comment by jvatic** [comment]
@titanous comments addressed

File: controller/grpc/controller.proto [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+  uint32 gid = 8;
+  // Write the file to the allowed list
+  bool allow = 9;
+}
+
+// See github.com/flynn/flynn/logaggregator/client Message
+message LogChunk {
+  // host_id is the host that the job was running on when this log message was
+  // emitted.
+  string host_id = 1;
```
Comment (jvatic) [inline controller/grpc/controller.proto]: Should this be a reference to a host resource? (e.g. `hosts/HOST_ID`)

File: controller/grpc/controller.proto [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+  // job_id is the ID of the job that emitted this log message.
+  string job_id = 2;
```
Comment (jvatic) [inline controller/grpc/controller.proto]: This should reference the job resource?

File: controller/grpc/controller.proto [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+message Deployment {
+  // name = "apps/APP_ID/deployments/DEPLOYMENT_ID"
+  string name = 1;
+  // old_release = Release.name
+  string old_release = 3;
+  // new_release = Release.name
+  string new_release = 4;
+  string strategy = 5;
+  string status = 6;
```
Comment (titanous) [inline controller/grpc/controller.proto]: I think this should be an enum?

File: controller/grpc/controller.proto [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+  // "stderr".
+  string stream = 6;
+  // create_time is the time that this log line was emitted.
+  google.protobuf.Timestamp create_time = 7;
+}
+
+// See github.com/flynn/flynn/logaggregator/types LogOpts
+message LogAggregatorLogOpts {
+  bool follow = 1;
+  string job_id = 2;
```
Comment (titanous) [inline controller/grpc/controller.proto]: `job` name?

File: controller/grpc/controller.proto [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+  // msg is the actual content of this log message.
+  string msg = 3;
+  // process_type is the type of process that emitted this log message.
+  string process_type = 4;
+  // source is the source of this log message, such as "app" or "router".
+  string source = 5;
+  // stream is the I/O stream that emitted this message, such as "stdout" or
+  // "stderr".
+  string stream = 6;
```
Comment (titanous) [inline controller/grpc/controller.proto]: This should probably be a `LogAggregatorStreamType`

File: controller/grpc/controller.proto (hunk 14 above)
Comment (titanous) [reply controller/grpc/controller.proto]: Yeah, just `job`?

File: controller/grpc/controller.proto (hunk 13 above)
Comment (titanous) [reply controller/grpc/controller.proto]: Makes sense.

File: controller/grpc/controller.proto (hunk 15 above)
Comment (jvatic) [reply controller/grpc/controller.proto]: It's not one in the controller, but looks like it should be. I'll create one for it here and make a note to refactor it into one in the controller.

File: controller/controller.go [hunk 18]
```diff
@@ -102,16 +106,42 @@ func main() {
...
+		cc:                 utils.ClusterClientWrapper(cluster.NewClient()),
+		lc:                 lc,
+		rc:                 rc,
+		defaultRouteDomain: defaultRouteDomain,
+		keys:               strings.Split(os.Getenv("AUTH_KEY"), ","),
+		keyIDs:             strings.Split(os.Getenv("AUTH_KEY_IDS"), ","),
+		caCert:             []byte(os.Getenv("CA_CERT")),
+	})
+
+	// TODO(jvatic): Set this up at the listener level
```
Comment (jvatic) [inline controller/controller.go]: Use [github.com/soheilhy/cmux](https://github.com/soheilhy/cmux)

**Pull request closed without merging**